Rewrite PO Token docs (#50)
yt-dlp/yt-dlp-wiki: PO Token Guide.md

There are two types of PO Tokens yt-dlp may require:
- GVS PO Token: This token is used on Google Video Server (GVS) requests.
- Player PO Token: This token is used on Innertube `player` requests.

These are not interchangeable, and you will need to provide both if you are using a client that requires both.

For more technical details on these tokens, refer to the [technical details section](#technical-details).

## Current PO Token enforcement

YouTube is at present rolling out changes to enforce PO Token requirements on clients. Currently, only GVS PO Tokens are required for some clients.

| Client         | GVS PO Token Required | Player PO Token Required | Notes                         |
|----------------|-----------------------|--------------------------|-------------------------------|
| `web`          | Yes                   | No                       |                               |
| `mweb`         | Yes                   | No                       |                               |
| `tv`           | No                    | No                       |                               |

> | `tv_embedded`  | No                    | No                       |                               |

Reviewer: ```suggestion
| `tv_embedded`  | No                    | No                       | Requires account cookies      |
```

| `web_embedded` | No                    | No                       |                               |
| `web_music`    | Yes                   | No                       |                               |
| `web_creator`  | Yes                   | No                       | Requires account cookies      |
| `android`      | Yes                   | ?                        | Account cookies not supported |
| `android_vr`   | No                    | No                       |                               |
| `ios`          | Yes                   | ?                        | Account cookies not supported |

> Note: HLS Formats do not require PO Token at this time.

Reviewer: ```suggestion
Note: HLS (m3u8) formats do not require PO Token at this time.
```

      * [No account](#no-account)
      * [With an account](#with-an-account)
    * [PO Token for Player](#po-token-for-player)
  * [Plugins](#plugins)
  * [Other PO Token related tools](#other-po-token-related-tools)
  * [Technical details](#technical-details)
    * [PO Tokens for GVS](#po-tokens-for-gvs)
    * [PO Tokens for Player](#po-tokens-for-player)
<!-- TOC -->

## Introduction

Proof of Origin (PO) Token is a parameter that YouTube requires to be sent with requests from some clients. Without it, requests for the affected clients' format URLs may return HTTP Error 403, or result in your account or IP address being blocked.

A PO Token is generated by an attestation provider on Web, Android and iOS platforms to attest the requests are coming from a genuine client.

For yt-dlp, you can provide PO Token(s) using the [`po_token` extractor argument](https://github.com/yt-dlp/yt-dlp#youtube), or you can [install a plugin](#plugins) that hooks into the YouTube extractor to handle fetching a PO Token.

For more technical details on these tokens, refer to the [technical details section](#technical-details).

## Cases where a PO Token is required

> There are currently two cases yt-dlp may require PO Tokens for video downloads, [depending on the client used](#current-po-token-enforcement):

Reviewer 2: In my opinion this should be moved down to the technical section (although the `Current PO Token enforcement` table shouldn’t be moved)

Author: This is needed to understand the manual extraction

Reviewer 2: Sorry, I hadn't read that far yet

You can select what client to use with the [`player_client` extractor argument](https://github.com/yt-dlp/yt-dlp#youtube).

## Providing a PO Token manually (for web client)

This section provides a basic guide on extracting PO Token(s) manually from YouTube in a web browser **for use with the `web` client**, and manually passing it to yt-dlp via the [`po_token` extractor argument](https://github.com/yt-dlp/yt-dlp#youtube). The same PO Token extraction method may work with other web browser-based clients too.

> [!TIP]
> When supplying multiple PO Tokens, use the same extractor args option and comma-separate the PO Token configurations. For example:
>   `--extractor-args "youtube:po_token=web.gvs+GVS_PO_TOKEN_VALUE_HERE,web.player+PLAYER_PO_TOKEN_VALUE_HERE"`

### PO Token for GVS

The PO Token used for `web` GVS requests is tied to your YouTube session. It generated differently depending on if you are logged in to yt-dlp or not.

#### No account

1. Open [YouTube Music](https://music.youtube.com) or YouTube Embedded (e.g. https://www.youtube.com/embed/aqz-KE-bpKQ) in a browser.  **Make sure you are not logged in to any account!**

> 2. Open any video

Reviewer 2: I'm pretty sure Github automatically handles lines starting with the wrong numbers. For maintenance purposes, I'd suggest changing every line to start with `1.`

Reviewer 2: Oh, I didn't notice that line 97 relies on knowing how many steps the PoToken guide is. That's gonna be fun to maintain **sigh**

Author: Not hard to maintain. IDE/linter complains if numbering is off

> 3. Open the developer console (F12), then go to the "Network" tab and filter by `googlevideo`

Reviewer 2: ```suggestion
1. Open the developer console (F12), then open the "Network" tab (you may need to first click >>)
1. In the search box, type `googlevideo` 
```

IMO >> is close enough to what the Chrome symbol looks like, unless anyone has any better ideas

Author: added

| `web_music`    | Yes                       | No                           |                                       |
| `web_creator`  | Yes                       | No                           | Requires account cookies              |
| `android`      | Yes                       | ?                            | Account cookies not supported         |
| `android_vr`   | No                        | No                           | YouTube Kids videos are not available |
| `ios`          | Yes*                      | ?                            | Account cookies not supported         |

*Client provides HLS (m3u8) formats which do not require PO Token for GVS at this time.

You can select what client to use with the [`player_client` extractor argument](https://github.com/yt-dlp/yt-dlp#youtube).

## Guide: Providing a PO Token manually (for use with `web` client)

This section provides a basic guide on extracting PO Token(s) manually from YouTube in a web browser **for use with the `web` client**, and manually passing it to yt-dlp via the [`po_token` extractor argument](https://github.com/yt-dlp/yt-dlp#youtube). 

> The same PO Token extraction method _may_ work with other web browser-based clients too.

Reviewer: do we want to spell this out here since it's gonna be a common use case
```suggestion
The same PO Token extraction method _may_ work with other web browser-based clients too (such as `web_music`).
```

Author: not really a fan. web_music currently has different Player PO Token for instance. Not necessary the same, and prone to change

### PO Token for GVS

The PO Token used for `web` GVS requests is tied to your YouTube session. It generated differently depending on if you are logged in to yt-dlp or not.

#### No account

1. Open [YouTube Music](https://music.youtube.com) or YouTube Embedded (e.g. https://www.youtube.com/embed/aqz-KE-bpKQ) in a browser.  **Make sure you are not logged in to any account!**
2. Open any video
3. Open the developer console (F12), then go to the "Network" tab (click the `>>` button if you don't see it)
4. Filter requests by `googlevideo.com`
5. Click the video and play for a few seconds - requests to `googlevideo.com` should appear in the network tab
6. From the most recent `googlevideo.com` request, extract the `pot` query parameter value from the URL
7. Pass the PO Token for GVS to yt-dlp using `--extractor-args "youtube:po_token=web.gvs+PO_TOKEN_VALUE_HERE"` with cookies (`--cookies COOKIES_FILE` or `--cookies-from-browser`)

Although not recommended, you may also provide visitor data instead of cookies. Refer to [Passing Visitor Data without cookies](https://github.com/yt-dlp/yt-dlp/wiki/Extractors#passing-visitor-data-without-cookies).

> 

Reviewer: where do we want to note the necessary `player_skip=webpage,configs` arg? here or in the main youtube section?

Author: added back, was accidentality removed
